sdks/python/apache_beam/runners/portability/spark_uber_jar_job_server.py

```diff
@@ -0,0 +1,318 @@
+
+import itertools
+import json
+import logging
+import shutil
+import tempfile
+import time
+import zipfile
+from concurrent import futures
+
+import grpc
+import requests
+from google.protobuf import json_format
+
+from apache_beam.options import pipeline_options
+from apache_beam.portability.api import beam_artifact_api_pb2_grpc
+from apache_beam.portability.api import beam_job_api_pb2
+from apache_beam.portability.api import endpoints_pb2
+from apache_beam.runners.portability import abstract_job_service
+from apache_beam.runners.portability import artifact_service
+from apache_beam.runners.portability import job_server
+from apache_beam.utils.timestamp import Timestamp
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class SparkUberJarJobServer(abstract_job_service.AbstractJobServiceServicer):
+  """A Job server which submits a self-contained Jar to a Spark cluster.
+
+  The jar contains the Beam pipeline definition, dependencies, and
+  the pipeline artifacts.
+  """
+
+  def __init__(self, master_url, options):
+    super(SparkUberJarJobServer, self).__init__()
+    self._master_url = master_url
+    self._executable_jar = (options.view_as(pipeline_options.SparkRunnerOptions)
+                            .spark_job_server_jar)
+    self._artifact_port = (options.view_as(pipeline_options.JobServerOptions)
+                           .artifact_port)
+    self._temp_dir = tempfile.mkdtemp(prefix='apache-beam-spark')
+
+  def start(self):
+    return self
+
+  def stop(self):
+    pass
+
+  def executable_jar(self):
+    url = (self._executable_jar or
+           job_server.JavaJarJobServer.path_to_beam_jar(
+               'runners:spark:job-server:shadowJar'))
+    return job_server.JavaJarJobServer.local_jar(url)
+
+  def create_beam_job(self, job_id, job_name, pipeline, options):
+    return SparkBeamJob(
+        self._master_url,
+        self.executable_jar(),
+        job_id,
+        job_name,
+        pipeline,
+        options,
+        artifact_port=self._artifact_port)
+
+
+class SparkBeamJob(abstract_job_service.AbstractBeamJob):
+  """Runs a single Beam job on Spark by staging all contents into a Jar
+  and uploading it via the Spark Rest API.
+
+  Note that the Spark Rest API is not enabled by default. It must be enabled by
+  setting the configuration property spark.master.rest.enabled to true."""
+
+  # These must agree with those defined in PortablePipelineJarUtils.java.
+  PIPELINE_FOLDER = 'BEAM-PIPELINE'
+  PIPELINE_MANIFEST = PIPELINE_FOLDER + '/pipeline-manifest.json'
+
+  # We only stage a single pipeline in the jar.
+  PIPELINE_NAME = 'pipeline'
+  PIPELINE_PATH = '/'.join(
+      [PIPELINE_FOLDER, PIPELINE_NAME, "pipeline.json"])
+  PIPELINE_OPTIONS_PATH = '/'.join(
+      [PIPELINE_FOLDER, PIPELINE_NAME, 'pipeline-options.json'])
+  ARTIFACT_MANIFEST_PATH = '/'.join(
+      [PIPELINE_FOLDER, PIPELINE_NAME, 'artifact-manifest.json'])
+  ARTIFACT_FOLDER = '/'.join([PIPELINE_FOLDER, PIPELINE_NAME, 'artifacts'])
```

Comment:
This and everything below do not look Spark-specific. Should we create a base class for ths jar functionality?

Reply:
Done.